File: src/main/java/org/nyxlang/interpreter/visitor/VarAccessVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.VarAccessNode;


/**
 * Interprets the var access node.
 */
public class VarAccessVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return VarAccessNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var varAccessNode = (VarAccessNode) node;
        var result = interpreter.getGlobalMemory().get(varAccessNode.getIdentifier());
        if (result == null) {
            throw new VisitorException("Variable with name " + varAccessNode.getIdentifier() + " not declared");
        }
        return result;
    }
}

File: src/main/kotlin/org/nyxlang/interpreter/visitor/VarAccessVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.interpreter.exception.VisitorException
import org.nyxlang.interpreter.result.DataRuntimeResult
import org.nyxlang.interpreter.result.dataResult
import org.nyxlang.parser.ast.INode
import org.nyxlang.parser.ast.AccessNode

/**
 * Interprets the var access node.
 */
class VarAccessVisitor : IVisitor {
    override fun visit(interpreter: IInterpreter, node: INode): DataRuntimeResult {
        val varAccessNode = node as AccessNode
        val varName = varAccessNode.name
        val value = interpreter.activationRecord!![varName]
                ?: throw VisitorException("Variable with name $varName undefined")
        return dataResult(value)
    }
}